Sources/ScheduledTasks.php

```diff
@@ -1542,4 +1542,178 @@ function scheduled_remove_old_drafts()
+				SELECT DISTINCT id_member
+				FROM {db_prefix}log_boards
+			) lb ON lb.id_member = m.id_member
+			WHERE m.last_login <= {int:dcutoff}
+		UNION
+		SELECT lmr.id_member, m.last_login
+			FROM {db_prefix}members m
+			INNER JOIN
+			(
+				SELECT DISTINCT id_member
+				FROM {db_prefix}log_mark_read
+			) lmr ON lmr.id_member = m.id_member
+			WHERE m.last_login <= {int:dcutoff}
+		UNION
+		SELECT lt.id_member, m.last_login
+			FROM {db_prefix}members m
+			INNER JOIN
+			(
+				SELECT DISTINCT id_member
+				FROM {db_prefix}log_topics
+				WHERE unwatched = {int:unwatched}
+			) lt ON lt.id_member = m.id_member
+			WHERE m.last_login <= {int:mrcutoff}
+		ORDER BY last_login
+		LIMIT {int:limit}';
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'dcutoff' => $cleanupBeyond,
+			'mrcutoff' => $markReadCutoff,
+			'unwatched' => 0,
+		)
+	);
+
+	// Move to array...
+	while($row = $smcFunc['db_fetch_assoc']($result))
+		$members[$row['id_member']] = $row['last_login'];
+	$smcFunc['db_free_result']($result);
+
+	// Nothing to do?
+	if (empty($members))
+		return true;
+
+	// Determine action based on last_login...
+	$purgeMembers = array();
+	$markReadMembers = array();
+	foreach($members as $id => $last_login)
+	{
+		if ($last_login <= $cleanupBeyond)
+			$purgeMembers[] = $id;
+		elseif ($last_login <= $markReadCutoff)
+			$markReadMembers[] = $id;
+	}
+
+	if (!empty($purgeMembers))
+	{
+		// Delete rows from log_boards
+		$sql = 'DELETE FROM {db_prefix}log_boards
+			WHERE id_member IN ({array_int:members})';
+		$smcFunc['db_query']('', $sql,
+			array(
+				'members' => $purgeMembers,
+			)
+		);
+		// Delete rows from log_mark_read
+		$sql = 'DELETE FROM {db_prefix}log_mark_read
+			WHERE id_member IN ({array_int:members})';
+		$smcFunc['db_query']('', $sql,
+			array(
+				'members' => $purgeMembers,
+			)
+		);
+		// Delete rows from log_topics
+		$sql = 'DELETE FROM {db_prefix}log_topics
+			WHERE id_member IN ({array_int:members})
+				AND unwatched = {int:unwatched}';
+		$smcFunc['db_query']('', $sql,
+			array(
+				'members' => $purgeMembers,
+				'unwatched' => 0,
+			)
+		);
+	}
+
+	// Nothing left to do?
+	if (empty($markReadMembers))
+		return true;
+
+	// Find board inserts to perform...
+	$boards = array();
+	foreach($markReadMembers as $id)
+	{
+		// Get board info for each member from log_topics.
+		// Note this user may have read many topics on that board, 
+		// but we just want one row each, & the ID of the last message read in each board.
+		$sql = 'SELECT lt.id_member, t.id_board, MAX(lt.id_msg) AS id_last_message
+			FROM {db_prefix}topics t
+			INNER JOIN
+			(
+				SELECT id_member, id_topic, id_msg
+				FROM {db_prefix}log_topics
+				WHERE id_member = {int:member}
+			) lt ON lt.id_topic = t.id_topic
+			GROUP BY lt.id_member, t.id_board';
+		$result = $smcFunc['db_query']('', $sql,
+			array(
+				'member' => $id,
+			)
+		);
+		while($row = $smcFunc['db_fetch_assoc']($result))
```

Comment:
maybe you want to use `` $smcFunc['db_fetch_all']($result) `` instead of this while loop.

Reply:
I tried both suggestions.  Slightly more readable & slight performance improvements.